### PR TITLE
[TRTLLM-14778][perf] Cache the Whisper suppress-token index as a device tensor

### PR DESCRIPTION
**Pull request by pranav-nvidia**
## Description

`_WhisperSuppressTokensLogitsProcessor` masks suppressed tokens with
`target[..., self.suppress_token_ids] = -inf`, where `suppress_token_ids` is a Python list of ~90
ids taken from the HF generation config. Indexing a CUDA tensor with a Python sequence builds a CPU
index tensor and blocking-copies it to the device — and that happens once per request per decode
step, so a batch of 32 over ~94 decoder steps performs roughly 3000 blocking host-to-device copies
per iteration.

The cost is not ordinary host work. A blocking copy serializes the decode loop, so it stays on the
critical path regardless of how much GPU work is available to overlap with it, and it is not hidden
by the overlap scheduler.

This caches the index as a device tensor, one entry per device the processor is used on, and masks
with `index_fill_`. The suppress id sequences become tuples so the cache can be keyed on device
alone: they cannot be mutated in place, so a cached index can never fall out of step with the ids it
was built from. The tuples are internal to this private class — nothing outside `llm.py` reads them.

Measured on whisper-tiny fp16, B200, 32 LibriSpeech clips, interleaved paired A/B (6 rounds,
baseline and change back to back within each round), overlap scheduler on, encoder and mixed
enc/dec CUDA graphs on:

| batch size | baseline RTFx | with change | delta | rounds won |
|---|---:|---:|---:|---|
| 8 | 622.06 | 781.05 | **+24.9 %** (95 % CI [+21.5, +27.7]) | 6/6 |
| 32 | 986.81 | 1223.95 | **+24.2 %** (95 % CI [+22.9, +24.9]) | 6/6 |

WER is unchanged at 0.1250 over the 32 clips in every arm, and the decoder step count per batch is
identical, so the speedup does not come from generating different output.

The same call site also performs an `inspect.signature` lookup per request per decode step. That is
a separate concern on the generic logits-processor path rather than a Whisper one, so it is left for
a follow-up PR; measured independently on the same setup it is worth a further +5.6 % / +7.5 %.

## Test Coverage

`tests/unittest/llmapi/test_whisper_suppress_tokens_processor.py` (new, `cpu_only`, 7 cases). The
file deliberately mixes two kinds of test:

Behavioural equivalence — these pass both with and without the production change, which is what
makes them useful as a guard on the masking semantics:
- `test_suppress_masks_exactly_the_listed_tokens`
- `test_begin_suppress_applies_only_while_the_window_is_open`
- `test_masks_each_beam_independently`
- `test_out_of_range_token_id_raises_every_call` — caching must not turn a hard error into a
  one-shot one

Guarding the change itself — these fail if the production change is reverted, verified by running
them against the pristine `main` version of `llm.py`:
- `test_token_ids_are_immutable`
- `test_index_tensor_is_built_once_per_device`
- `test_empty_lists_mask_nothing`

Against `origin/main`: 3 failed, 4 passed. With the change: 7 passed.

Existing coverage that exercises this path end to end:
`tests/integration/defs/llmapi/test_llm_api_pytorch_whisper.py` (exact-output checks over the
feature-combination matrix).

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Dev Engineer Review

- `_WhisperSuppressTokensLogitsProcessor` stores suppression IDs as immutable tuples.
- Per-device tensor caches avoid Python-list indexing on CUDA tensors.
- `index_fill_` applies suppression masks efficiently.
- Cache state is excluded from copying and serialization.
- Cache initialization is safe for concurrent access.
- The implementation handles empty IDs, out-of-range IDs, and separate device caches.
- Test-list entries use valid paths and the intended A10 and CPU scope.
- No configuration or test-list format issues were identified.

## QA Engineer Review

- Added tests for masking, prompt-window handling, beam independence, immutable IDs, empty inputs, out-of-range IDs, per-device caches, CUDA placement, half precision, copying, pickling, and concurrent cache initialization.
- Added the test to:
  - `tests/integration/test_lists/test-db/l0_a10.yml`
  - `tests/integration/test_lists/test-db/l0_cpu.yml`
- The test is covered in CI through both test lists.
- Verdict: sufficient.

### CODE REVIEW

File: tests/unittest/llmapi/test_whisper_suppress_tokens_processor.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+    both_missed = threading.Barrier(2, timeout=30)
+    real_tensor = torch.tensor
+    builds = []
+
+    def rendezvous_tensor(*args, **kwargs):
+        builds.append(None)
+        both_missed.wait()
+        return real_tensor(*args, **kwargs)
+
+    monkeypatch.setattr(torch, "tensor", rendezvous_tensor)
```
Comment (brnguyen2) [inline tests/unittest/llmapi/test_whisper_suppress_tokens_processor.py:209]: `torch.tensor` is patched process-wide while both threads sit in a 2-party barrier with a 30 s timeout. Any unrelated `torch.tensor` call in the process during that window (another thread, a torch-internal helper) becomes a third waiter: it opens a new barrier generation, blocks 30 s, and raises `BrokenBarrierError` from somewhere unrelated to this test. Low probability today, but a hard-to-diagnose flake if it ever fires in CI.

Cheapest fix that keeps the contention you want: have `rendezvous_tensor` rendezvous only for the first two calls (release immediately once `len(builds) > 2`), or gate on the calling thread being one of the two started here.

Comment (pranav-nvidia) [reply tests/unittest/llmapi/test_whisper_suppress_tokens_processor.py:209]: Fixed in `60ee446`, I went with the thread-identity gate. The `len(builds) > 2` approach would miss a stray thread calling `torch.tensor` and cause a silent failure rather than a `BrokenBarrierError`. 

`rendezvous_tensor` now returns the real `torch.tensor` immediately for any thread outside the two started here.

Also tightened the joins below in `c049704`.

**Review by lori-ren — APPROVED** [review]

File: tests/unittest/llmapi/test_whisper_suppress_tokens_processor.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+import torch
+
+from tensorrt_llm.llmapi.llm import _WhisperSuppressTokensLogitsProcessor
+
+# `cpu_only` is applied per test rather than to the module. GPU stages select with
+# `-m "not cpu_only"` (jenkins/L0_Test.groovy), so a module-level mark would deselect the
+# whole file there while the CUDA cases below skip on CPU stages for want of a device -
+# leaving them running nowhere. Host-only tests carry the mark; the CUDA ones do not.
+cpu_only = pytest.mark.cpu_only
+requires_cuda = pytest.mark.skipif(not torch.cuda.is_available(), reason="requires CUDA")
```
[the remark this answers is not in the document]
Comment (pranav-nvidia) [reply tests/unittest/llmapi/test_whisper_suppress_tokens_processor.py:32]: @coderabbitai UPPER_SNAKE_CASE is correct in general, but repo convention for pytest marker aliases is 74 lowercase / 0 uppercase across tests/ — including 7 in the shared tests/unittest/utils/util.py (skip_pre_hopper, skip_ray, …). Renaming would make this file the only outlier. I'm declining this suggestion.
